<a href="https://colab.research.google.com/github/neilsambhu/carla-simulator2/blob/main/carla_simulator2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Sandbox for CARLA simulator following https://github.com/MichaelBosello/carla-colab. Original *.ipynb file: https://colab.research.google.com/github/MichaelBosello/carla-colab/blob/master/carla-simulator.ipynb. 

In case of poor functionality: Runtime > Restart runtime (Ctrl+M.)

2/17/2022 8:49 PM: instead of (1) "!" to execute shell commands, use (2) the Python "subprocess" module to ensure compatibility between iPython and Python (complete as of 2/17/2022 10:16 PM).

2/17/2022 10:23 PM: I made the Call_SubprocessPopen(cmd) function for easily calling shell commands. 

2/17/2022 11:02 PM: The subprocess.run will fail for command-line arguments separated by spaces. The subprocess.run requires a list.

In [1]:
import subprocess, sys
from sys import platform
def Call_SubprocessPopen(cmd):
  p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
  # p = subprocess.run(cmd, capture_output=True, text=True, check=True)
  for line in p.stdout.readlines():
    sys.stdout.write(line.decode("utf-8"))
    # sys.stdout.write((r"utf-8")line)
    pass
  # print(p.stdout)
if platform == "linux" or platform == "linux2":
  # for multi-line output: use "pip --help" instead of "pwd" 
  cmd = "pwd"
  # cmd = "pip --help"
  Call_SubprocessPopen(cmd)
elif platform == "win32":
  raise Exception("2/17/2022 8:59 PM: This code is designed to run in a Linux-based environment.")

/content


In [2]:
# <10 seconds
Call_SubprocessPopen("pip install git+https://github.com/demotomohiro/remocolab.git")

  Cloning https://github.com/demotomohiro/remocolab.git to /tmp/pip-req-build-haa52oex
  Running command git clone -q https://github.com/demotomohiro/remocolab.git /tmp/pip-req-build-haa52oex
  Created wheel for remocolab.py: filename=remocolab.py-0.1-py3-none-any.whl size=7553 sha256=07b1e2ecaaa4c37c90cff7b7b80c959aef2e8b5899852e08302f19e2f095a807
  Stored in directory: /tmp/pip-ephem-wheel-cache-x6a2m41_/wheels/80/37/44/386e71ab0cd5aef5cefdb1926b710a42fdb4fe2d07c3b1b08f
  Created wheel for pyngrok: filename=pyngrok-5.1.0-py3-none-any.whl size=19007 sha256=b2e28b67ed9dbc7a5c14dae287b55ae9a8b1be9791d9ea784681918a6c4517e1
  Stored in directory: /root/.cache/pip/wheels/bf/e6/af/ccf6598ecefecd44104069371795cb9b3afbcd16987f6ccfb3
Successfully built remocolab.py pyngrok


In [3]:
# <5.5 minutes
import remocolab
remocolab.setupVNC()

Warning! GPU of your assigned virtual machine is Tesla K80.
You might get better GPU by reseting the runtime.
Do you want to continue? [y/n] y
---
As ngrok doesn't work on colab for long time, default tunnel method has been changed to Argo tunnel.
Please read this for more details:
https://github.com/demotomohiro/remocolab/blob/master/README.md


Label(value='')

FloatProgress(value=0.0, layout=Layout(border='1px solid #118800'), max=1.0)

Install openssh-server


Label(value='')

FloatProgress(value=0.0, layout=Layout(border='1px solid #118800'), max=1.0)

Install xfce4
Install xfce4-terminal
ED25519 key fingerprint of host:
256 SHA256:WhMoMGN+I0xkQvMXr1He52E9FoAomKSnZ9emcxuASR0 root@45c5895d295f (ED25519)
+--[ED25519 256]--+
|o+X.oE .. ....   |
| B+*..*o..  . .  |
| .++++oo.. + +   |
|  +o+o+  .+ o .  |
| . = + oS  .     |
|  o . +o .       |
|     o.o         |
|      o o        |
|       .         |
+----[SHA256]-----+

✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
root password: 5Ni8V3AsehHZsY-6CV61HuBv6R89zcZ6-8tza7qigUo
colab password: HEpZXYn7N1tM-crkXa5sdSqijiPWWgrnnWOEwMLpWoQ
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
---
Execute following command on your local machine and login before running TurboVNC viewer:
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -oProxyCommand="cloudflared access ssh --hostname %h" -L 5901:localhost:5901 colab@vat-bracket-mean-hobbies.trycloudflare.com
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
VNC password: Fc8esAoz
VNC view on

2/20/2022 9:01 PM: Problem:
ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -oProxyCommand="cloudflared access ssh --hostname %h" -L 5901:localhost:5901 colab@vat-bracket-mean-hobbies.trycloudflare.com
/bin/bash: line 0: exec: cloudflared: not found
kex_exchange_identification: Connection closed by remote host

[install cloudflared (https://pkg.cloudflare.com/#ubuntu-title) in Windows Subsystem for Linux]

2/20/2022 9:08 PM: Problem
ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -o ProxyCommand="cloudflared access ssh --hostname %h" -L 5901:localhost:5901 colab@vat-bracket-mean-hobbies.trycloudflare.com
/bin/bash: line 0: exec: cloudflared: not found
kex_exchange_identification: Connection closed by remote host